lib/matplotlib/pyplot.py

```diff
@@ -2180,9 +2180,13 @@ def set_cmap(cmap):
     matplotlib.cm.register_cmap
     matplotlib.cm.get_cmap
     """
+    # The name of the colormap may be different than the registered name.
+    # We want the registered name for the rc setting
+    name = getattr(cmap, "name", cmap)
+    # Run the cmap (str or Colormap) through get_cmap to validate
+    # the cmap before updating the rc parameters
     cmap = get_cmap(cmap)
-
-    rc('image', cmap=cmap.name)
+    rc('image', cmap=name)
```

Comment:
If I do:

```python
my_cmap = mpl.colormaps["viridis"].copy()
my_cmap.set_bad("FF00FF")
mpl.colormaps.register("my-cmap", my_cmap) # note, did not update cmap.name
plt.set_cmap(my_cmap)  # note, passing instance, not name
```

Won't this get `"viridis"` from the `getattr`, and thus set that as the rc cmap? (It will apply the "correct" cmap to `gci`, but not to future mappables)

Not sure how to square that without a) doing a reverse lookup in the registry or b) requiring that the names match when registered.

Perhaps this change is still better than no change, but not sure it's without its rough edges still... certainly it makes it possible to use a mismatched cmap, but you do need to get it by the registry name, not passing the cmap object.